presto-main/src/main/java/com/facebook/presto/operator/aggregation/arrayagg/ArrayAggregationFunction.java

```diff
@@ -137,7 +141,13 @@ public static void output(Type elementType, ArrayAggregationState state, BlockBu
         }
         else {
             BlockBuilder entryBuilder = out.beginBlockEntry();
-            state.forEach((block, position) -> elementType.appendTo(block, position, entryBuilder));
+            state.forEach((block, position) -> {
+                if (entryBuilder.getRetainedSizeInBytes() > MAX_ARRAY_SIZE) {
```

Comment:
I think you might want to "cut the line" earlier. For example, if the current retained size is already `MAX_ARRAY_SIZE - 10`, and the next element has length 11 (thinking about `ARRAY<VARCHAR>`). You will still get cannot allocate Slice XXX error.

I would say just define an constant like `MAX_ENTRY_RETAINED_SIZE = 1_000_000_000` in this class.

Reply:
Also, does it make sense to do the size check after the `appendTo` call? :)

Reply:
if we're cutting the line at almost half the actual limit I don't see much difference between before or after.. Earlier I was erring towards new error as opposed to ```INTERNAL_ERROR``` coming from airlift. Leaving it like this but happy to discuss if you feel strongly about moving the check to after the append

Reply:
@sujay-jain : I usually don't click "Resolve conversation" if it's still an open conversation :). I will leave it to the commenter to resolve it .

It's not about the behavior, I am pretty sure in almost real cases, they have the same result. It's from code flow perspective, you want to do the size check after you append something, rather than before you don't do any change ( thinking about eager check vs. lazy check). And in genenral we prefer eager check so the exception happened right after the problem happens.